src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContextSerializer.cs

```diff
@@ -25,25 +25,165 @@ private DesigntimeLicenseContextSerializer()
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                Debug.Assert(_stream.Position != 0, "Expected the first byte to be read first");
+                if (_stream.Position == 1)
+                {
+                    Debug.Assert(firstSByte != -100);
+                    // Add the first byte read by ReadByte into buffer here
+                    buffer[0] = (byte)firstSByte;
+                    return _stream.Read(buffer, offset + 1, count - 1) + 1;
+                }
+                return _stream.Read(buffer, offset, count);
+            }
+
+            public override long Seek(long offset, SeekOrigin origin) => _stream.Seek(offset, origin);
+
+            public override void SetLength(long value) => _stream.SetLength(value);
+
+            public override void Write(byte[] buffer, int offset, int count) => _stream.Write(buffer, offset, count);
+
+            public override int ReadByte()
+            {
+                sbyte read = (sbyte)_stream.ReadByte();
+                firstSByte = read;
+                return read;
+            }
+        }
+
+        /// <summary>
+        /// During deserialization, the stream passed in may be binary formatted or may have used binary writer. This is quick test to discern between them.
+        /// </summary>
+        internal static bool StreamIsBinaryFormatted(StreamWrapper stream)
+        {
+            // For binary formatter, the first byte is the SerializationHeaderRecord and has a value 0
+            int firstByte = stream.ReadByte();
+            if (firstByte != 0)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static void DeserializeUsingBinaryFormatter(StreamWrapper wrappedStream, string cryptoKey, RuntimeLicenseContext context)
         {
+            try
+            {
 #pragma warning disable SYSLIB0011 // Issue https://github.com/dotnet/runtime/issues/39293 tracks finding an alternative to BinaryFormatter
-            IFormatter formatter = new BinaryFormatter();
+                IFormatter formatter = new BinaryFormatter();
 
-            object obj = formatter.Deserialize(o);
+                object obj = formatter.Deserialize(wrappedStream);
 #pragma warning restore SYSLIB0011
 
-            if (obj is object[] value)
+                if (obj is object[] value)
+                {
+                    if (value[0] is string && (string)value[0] == cryptoKey)
+                    {
+                        context._savedLicenseKeys = (Hashtable)value[1];
+                    }
+                }
+            }
+            catch (NotSupportedException exception)
+            {
+                if (!LocalAppContextSwitches.BinaryFormatterSerializationEnabled)
+                {
+                    throw new NotSupportedException(exception.Message + " Turn on the EnableUnsafeBinaryFormatterSerialization flag to continue using BinaryFormatter");
```

Comment:
Do we want to put this message in a .resx?

Reply:
Yup. I'm just waiting to see if @GrabYourPitchforks wants to modify `exception.Message` to include a blurb about the `EnableUnsafeBinaryFormatterSerialization` flag.